lib/collection/src/collection_manager/optimizers/config_mismatch_optimizer.rs

```diff
@@ -0,0 +1,481 @@
+
+/// Looks for segments having a mismatch between configured and actual parameters
+///
+/// For example, a user may change the HNSW parameters for a collection. A segment that was already
+/// indexed with different parameters now has a mismatch. This segment should be optimized (and
+/// indexed) again in order to update the effective configuration.
+pub struct ConfigMismatchOptimizer {
+    thresholds_config: OptimizerThresholds,
+    segments_path: PathBuf,
+    collection_temp_dir: PathBuf,
+    collection_params: CollectionParams,
+    hnsw_config: HnswConfig,
+    quantization_config: Option<QuantizationConfig>,
+    telemetry_durations_aggregator: Arc<Mutex<OperationDurationsAggregator>>,
+}
+
+impl ConfigMismatchOptimizer {
+    pub fn new(
+        thresholds_config: OptimizerThresholds,
+        segments_path: PathBuf,
+        collection_temp_dir: PathBuf,
+        collection_params: CollectionParams,
+        hnsw_config: HnswConfig,
+        quantization_config: Option<QuantizationConfig>,
+    ) -> Self {
+        ConfigMismatchOptimizer {
+            thresholds_config,
+            segments_path,
+            collection_temp_dir,
+            collection_params,
+            hnsw_config,
+            quantization_config,
+            telemetry_durations_aggregator: OperationDurationsAggregator::new(),
+        }
+    }
+
+    fn worst_segment(
+        &self,
+        segments: LockedSegmentHolder,
+        excluded_ids: &HashSet<SegmentId>,
+    ) -> Vec<SegmentId> {
+        let segments_read_guard = segments.read();
+        let candidates: Vec<_> = segments_read_guard
+            .iter()
+            // Excluded externally, might already be scheduled for optimization
+            .filter(|(idx, _)| !excluded_ids.contains(idx))
+            .filter_map(|(idx, segment)| {
+                let segment_entry = segment.get();
+                let read_segment = segment_entry.read();
+                let point_count = read_segment.available_point_count();
+                let vector_size = point_count
+                    * read_segment
+                        .vector_dims()
+                        .values()
+                        .max()
+                        .copied()
+                        .unwrap_or(0)
+                    * VECTOR_ELEMENT_SIZE;
+
+                let segment_config = read_segment.config();
+
+                if read_segment.segment_type() == SegmentType::Special {
+                    return None; // Never optimize already optimized segment
+                }
+
+                // Determine whether segment has mismatch
+                let has_mismatch =
+                    segment_config
+                        .vector_data
+                        .iter()
+                        .any(|(vector_name, vector_data)| {
+                            // Check HNSW mismatch
+                            match &vector_data.index {
+                                Indexes::Plain {} => {}
```

Comment:
This `match` could be simplified a bit, IMO. (The following two comments/suggestions are also a part of the same "simplification".)

```suggestion
                                Indexes::Plain {} => false,
```